kstat.rs

```diff
@@ -0,0 +1,313 @@
+struct KstatWrapper {
+    kc: *mut KstatCtl,
+    ksp: *mut Kstat,
+    stepping: bool,
+}
+
+/// Turn an optional string reference into an optional owned C string.
+fn cstr(s: Option<&str>) -> Option<std::ffi::CString> {
+    s.map_or(None, |s| Some(CString::new(s.to_owned()).unwrap()))
+}
+
+/// Turn an optional C string into a (const char *) for Some, or NULL for None.
+fn cstrp(p: &Option<CString>) -> *const c_char {
+    if let Some(p) = p.as_ref() {
+        p.as_ptr()
+    } else {
+        null()
+    }
+}
+
+impl KstatWrapper {
+    fn open() -> Result<Self> {
+        let kc = unsafe { kstat_open() };
+        if kc == null_mut() {
+            return Err("kstat_open(3KSTAT) failed".into());
+        }
+        Ok(KstatWrapper {
+            kc: kc,
+            ksp: null_mut(),
+            stepping: false,
+        })
+    }
+
+    /// Call kstat_lookup(3KSTAT) and store the result, if there is a match.
+    fn lookup(&mut self, module: Option<&str>, name: Option<&str>) {
+        let module = cstr(module);
+        let name = cstr(name);
+
+        unsafe {
+            self.ksp = kstat_lookup(self.kc, cstrp(&module), -1, cstrp(&name));
+        }
+
+        self.stepping = false;
+    }
+
+    /// Call once to start iterating, and then repeatedly for each additional
+    /// kstat in the chain.  Returns false once there are no more kstat entries.
+    fn step(&mut self) -> bool {
+        if !self.stepping {
+            self.stepping = true;
+        } else {
+            self.ksp = unsafe { (*self.ksp).ks_next };
+        }
+
+        if self.ksp == null_mut() {
+            self.stepping = false;
+            false
+        } else {
+            true
+        }
+    }
+
+    /// Return the module name of the current kstat.
+    fn module(&self) -> Option<String> {
+        if self.ksp == null_mut() {
+            None
+        } else {
+            Some(unsafe { (*self.ksp).module() })
+        }
+    }
+
+    /// Return the name of the current kstat.
+    fn name(&self) -> Option<String> {
+        if self.ksp == null_mut() {
+            None
+        } else {
+            Some(unsafe { (*self.ksp).name() })
+        }
+    }
+
+    /// Look up a named kstat value.  For internal use by typed accessors.
+    fn data_value(&self, statistic: &str) -> Option<*const KstatNamed> {
+        if self.ksp == null_mut() {
+            return None;
+        }
+
+        if unsafe { kstat_read(self.kc, self.ksp, null_mut()) } == -1 {
+            return None;
+        }
+
+        let stat = cstr(Some(statistic));
+
+        unsafe {
+            let knp = kstat_data_lookup(self.ksp, cstrp(&stat));
+            if knp == null_mut() {
+                return None;
+            }
+
+            Some(knp as *const KstatNamed)
+        }
+    }
+
+    /// Look up a named kstat value and interpret it as a "long_t".
+    fn data_long(&self, statistic: &str) -> Option<i64> {
```

Comment:
If this is a C long, you probably want to use the `c_long` type instead of `i64`.

Reply:
It is a C `long`, and that type appears in `union KstatValue` above.  I'm casting to `i64` here (and `u64` for `data_ulong()` below) as that will provide a consistent size across both ILP32 and LP64 platforms.